aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -744,6 +746,72 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
 }
 
 DEFINE_DISPATCH(sparse_mask_intersection_out_stub);
+DEFINE_DISPATCH(sparse_mask_projection_out_stub);
+
+using OptTensor = c10::optional<Tensor>;
+
+std::tuple<Tensor, Tensor, OptTensor> sparse_mask_like_prepare_sparse_inputs(
+    const std::string& method_name,
+    const Tensor& t,
+    const Tensor& mask) {
+  // This is a helper function for operations that implement "sparse_mask"-like
+  // functionality, namely, projection of values of one tensor onto the other.
+  // These operations mostly rely on COO intersection primitives that heavily
+  // exploit coalesced inputs to avoid any syncs and calls to sort. The problem
+  // is that these primitives might project first argument onto second one or
+  // the other way around depending on which arguments are coalesced and which are
+  // larger. This function prepares inputs for `sparse_mask` such that `t` is
+  // projected onto `mask` by sorting `t` if uncoalesced and artifically marking it
+  // as coalesced all while `mask` is set to uncoalesced.
+  // The result of this projectionk is going to be uncoalesced, so it is up to the
+  // user to set the corresponding flag correctly with respect to the operations'
+  // semantics.
+
+  // We already assume that t.sizes() == mask.sizes()
+  TORCH_CHECK(t.sparse_dim() == mask.sparse_dim(),
+              method_name, "(): the number of sparse dimensions in `self` ",
+              "should match that of the `mask`. ",
+              "Got `self.sparse_dim() == ", t.sparse_dim(), "` != ",
+              "`mask.sparse_dim() == ", mask.sparse_dim(), "`.");
+
+  const auto wrapped_tensor = [](const Tensor& t,
+                                 const OptTensor& indices = c10::nullopt,
+                                 const OptTensor& values = c10::nullopt) -> Tensor {
+    auto res = at::empty({0}, t.options());
+    auto* res_sparse_impl = get_sparse_impl(res);
+    res_sparse_impl->raw_resize_(t.sparse_dim(), t.dense_dim(), t.sizes());
+    const auto res_indices = indices.has_value() ? *indices : t._indices();
+    const auto res_values = values.has_value() ? *values : t._values();
+    res_sparse_impl->set_indices_and_values_unsafe(res_indices, res_values);
+    res_sparse_impl->set_nnz_and_narrow(t._nnz());
```

Comment:
Is using `t._nnz()` safe considering that `res_values.size(0) == t._nnz()` is not always true?